pkg/registry/core/service/strategy.go

```diff
@@ -205,6 +207,61 @@ func topologyKeysInUse(svc *api.Service) bool {
 	return len(svc.Spec.TopologyKeys) > 0
 }
 
+// This is an unusual case.  Service has a number of inter-related fields and
+// in order to avoid breaking clients we try really hard to infer what users
+// mean when they change them.
+//
+// Services are effectively a discriminated union, where `type` is the
+// discriminator. Some fields just don't make sense with some types, so we
+// clear them.
+//
+// As a rule, we almost never change user input.  This can get tricky when APIs
+// evolve and new dependent fields are added.  This specific case includes
+// fields that are allocated from a pool and need to be released.  Anyone who
+// is contemplating copying this pattern should think REALLY hard about almost
+// any other option.
+func dropTypeDependentFields(newSvc *api.Service, oldSvc *api.Service) {
+	// For now we are only normalizing on updates.  This minimizes potential
+	// confusion since many of the cases we are handling here are pretty niche.
+	if oldSvc == nil {
+		return
+	}
+
+	// If the user is switching to a type that does not need a value (even
+	// "None" counts), we can wipe ClusterIP.  There are other fields that
+	// could also be wiped, but historically were ignored.  For compat, we
+	// leave it that way.
+	if needsClusterIP(oldSvc) && !needsClusterIP(newSvc) {
```

Comment:
also make sure newSvc did not change the clusterIP (if the incoming request explicitly monkeyed with the field, they should fail rather than get wiped and succeed)

Reply:
ACK, in next push